src/backend/common/middleware.py

```diff
@@ -66,6 +69,50 @@ def _run_after(self):
             execute_callbacks()
 
 
+def install_backend_security(app: Flask) -> None:
+    @app.before_request
+    def _confirm_security_headers():
+        # Only allow signed requests from TBA services, cron job, or task queue
+        task_header = request.headers.get("X-AppEngine-TaskName")
+        task_header_dev = request.headers.get("X-Google-TBA-RedisTask")
+        cron_header = request.headers.get("X-Appengine-Cron")
+        auth_header = request.headers.get("Authorization")
+        if task_header or task_header_dev or cron_header:
+            return
+        elif auth_header:
+            # Verify that a request is coming from another TBA service
+            # https://cloud.google.com/appengine/docs/standard/python/migrate-to-python3/migrate-app-identity
+
+            # The auth_header must be in the form Authorization: Bearer token.
+            bearer, token = auth_header.split()
+            if bearer.lower() != 'bearer':
+                return abort(403)
+
+            try:
+                info = id_token.verify_oauth2_token(token, requests.Request())
+                service_account_email = info['email']
+                incoming_app_id, domain = service_account_email.split('@')
+                if domain != 'appspot.gserviceaccount.com':  # Not App Engine svc acct
+                    return abort(403)
+
+                project = Environment.project()
+                if not project:
+                    return abort(403)
+
+                if incoming_app_id != project:
+                    return abort(403)
+
+                # TODO: Need to verify things here...
```

Comment:
- [ ] TODO: What does this even mean??

Reply:
- [ ] Test this with an upstream account. I was running in to issues using my upstream account - but the `test.json` key seems to work fine.

Reply:
For "what does this mean" - I'm pretty sure we just need to look at that info print and see if there's any other verifiable information we need here. Based on the docs linked, verifying the OAuth token, `incoming_app_id`, `domain`, and `project` should be enough. But let's double check, just in case.